cpp/src/stream_compaction/filter/filter.cu

```diff
@@ -0,0 +1,351 @@
+                                                      mr);
+}
+
+void launch_filter_kernel(jitify2::ConfiguredKernel& kernel,
+                          cudf::jit::device_span<cudf::size_type> output,
+                          std::vector<column_view> const& input_columns,
+                          std::optional<void*> user_data,
+                          rmm::cuda_stream_view stream,
+                          rmm::device_async_resource_ref mr)
+{
+  auto outputs = cudf::jit::to_device_vector(
+    std::vector{cudf::jit::device_optional_span<cudf::size_type>{output, nullptr}}, stream, mr);
+
+  auto [input_handles, inputs] =
+    cudf::jit::column_views_to_device<column_device_view, column_view>(input_columns, stream, mr);
+
+  cudf::jit::device_optional_span<cudf::size_type> const* outputs_ptr = outputs.data();
+  column_device_view const* inputs_ptr                                = inputs.data();
+  void* p_user_data                                                   = user_data.value_or(nullptr);
+
+  std::array<void*, 3> args{&outputs_ptr, &inputs_ptr, &p_user_data};
+
+  kernel->launch(args.data());
+}
+
+void perform_checks(column_view base_column,
+                    std::vector<column_view> const& columns,
+                    std::optional<std::vector<bool>> copy_mask)
+{
+  CUDF_EXPECTS(is_runtime_jit_supported(), "Runtime JIT is only supported on CUDA Runtime 11.5+");
+  CUDF_EXPECTS(std::all_of(columns.begin(),
+                           columns.end(),
+                           [](auto& input) {
+                             return is_fixed_width(input.type()) ||
+                                    (input.type().id() == type_id::STRING);
+                           }),
+               "Filters only support fixed-width and string types",
+               std::invalid_argument);
+
+  CUDF_EXPECTS(std::all_of(columns.begin(),
+                           columns.end(),
+                           [&](auto& input) {
+                             return cudf::jit::is_scalar(base_column.size(), input.size()) ||
+                                    (input.size() == base_column.size());
+                           }),
+               "All filter columns must have the same size or be scalar (have size 1)",
+               std::invalid_argument);
+
+  CUDF_EXPECTS(!copy_mask.has_value() || (copy_mask->size() == columns.size()),
+               "The size of the copy mask must match the number of input columns",
+               std::invalid_argument);
+}
+
+jitify2::Kernel get_kernel(std::string const& kernel_name, std::string const& cuda_source)
+{
+  return cudf::jit::get_program_cache(*stream_compaction_filter_jit_kernel_cu_jit)
+    .get_kernel(kernel_name,
+                {},
+                {{"cudf/detail/operation-udf.hpp", cuda_source}},
+                {"-arch=sm_.",
+                 "--device-int128",
+                 // TODO: remove when we upgrade to CCCL >= 3.0
```

Comment:
We are already using CCCL 3.0.

Reply:
(This can be deferred to a follow-up PR if you wish.)

Reply:
I've created an issue for it: https://github.com/rapidsai/cudf/issues/19344